File: rss.go
// rss.go
// RSS Reader (Фильтрация новостей) на Go

package main

import (
	"encoding/json"
	"encoding/xml"
	"flag"
	"fmt"
	"io/ioutil"
	"net/http"
	"os"
	"os/exec"
	"regexp"
	"strconv"
	"strings"
	"time"
)

// ANSI-цвета
const (
	reset  = "\033[0m"
	bold   = "\033[1m"
	red    = "\033[91m"
	green  = "\033[92m"
	yellow = "\033[93m"
	blue   = "\033[94m"
	magenta= "\033[95m"
	cyan   = "\033[96m"
	white  = "\033[97m"
	gray   = "\033[90m"
)

func colorize(text, color string) string {
	return color + text + reset
}

// Фильтр
type Filter struct {
	Keywords   []string `json:"keywords"`
	Days       int      `json:"days"`
	Source     string   `json:"source"`
	UnreadOnly bool     `json:"unread_only"`
	ReadOnly   bool     `json:"read_only"`
}

// Структуры RSS
type RSS struct {
	Channel struct {
		Title string `xml:"title"`
		Items []struct {
			Title       string   `xml:"title"`
			Link        string   `xml:"link"`
			PubDate     string   `xml:"pubDate"`
			Description string   `xml:"description"`
			Categories  []string `xml:"category"`
		} `xml:"item"`
	} `xml:"channel"`
}

type Atom struct {
	Title string `xml:"title"`
	Entries []struct {
		Title     string   `xml:"title"`
		Link      struct { Href string `xml:"href,attr"` } `xml:"link"`
		Published string   `xml:"published"`
		Updated   string   `xml:"updated"`
		Summary   string   `xml:"summary"`
		Content   string   `xml:"content"`
		Categories []struct { Term string `xml:"term,attr"` } `xml:"category"`
	} `xml:"entry"`
}

type FeedItem struct {
	ID          int      `json:"id"`
	Title       string   `json:"title"`
	Link        string   `json:"link"`
	PubDate     string   `json:"pubDate"`
	Description string   `json:"description"`
	Categories  []string `json:"categories"`
	Read        bool     `json:"read"`
}

type Feed struct {
	ID        int        `json:"id"`
	Title     string     `json:"title"`
	URL       string     `json:"url"`
	Items     []FeedItem `json:"items"`
	LastFetch string     `json:"last_fetch"`
}

type Data struct {
	Feeds      []Feed `json:"feeds"`
	NextFeedID int    `json:"next_feed_id"`
	NextItemID int    `json:"next_item_id"`
	Filter     Filter `json:"filter"`
}

type RssReader struct {
	dataFile string
	data     Data
	nextFeedID int
	nextItemID int
	filter   Filter
}

func NewRssReader(dataFile string) *RssReader {
	r := &RssReader{dataFile: dataFile}
	r.load()
	return r
}

func (r *RssReader) load() {
	content, err := ioutil.ReadFile(r.dataFile)
	if err != nil {
		r.data = Data{Feeds: []Feed{}, NextFeedID: 1, NextItemID: 1, Filter: Filter{}}
	} else {
		json.Unmarshal(content, &r.data)
	}
	r.nextFeedID = r.data.NextFeedID
	r.nextItemID = r.data.NextItemID
	r.filter = r.data.Filter
}

func (r *RssReader) save() {
	r.data.NextFeedID = r.nextFeedID
	r.data.NextItemID = r.nextItemID
	r.data.Filter = r.filter
	data, _ := json.MarshalIndent(r.data, "", "  ")
	ioutil.WriteFile(r.dataFile, data, 0644)
}

func (r *RssReader) fetchFeed(url string) (string, error) {
	resp, err := http.Get(url)
	if err != nil {
		return "", err
	}
	defer resp.Body.Close()
	body, err := ioutil.ReadAll(resp.Body)
	return string(body), err
}

func (r *RssReader) parseFeed(content string) (string, []FeedItem) {
	// Попробуем RSS
	var rss RSS
	if err := xml.Unmarshal([]byte(content), &rss); err == nil && rss.Channel.Title != "" {
		items := []FeedItem{}
		for _, it := range rss.Channel.Items {
			items = append(items, FeedItem{
				Title:       it.Title,
				Link:        it.Link,
				PubDate:     it.PubDate,
				Description: it.Description,
				Categories:  it.Categories,
			})
		}
		return rss.Channel.Title, items
	}
	// Попробуем Atom
	var atom Atom
	if err := xml.Unmarshal([]byte(content), &atom); err == nil && atom.Title != "" {
		items := []FeedItem{}
		for _, it := range atom.Entries {
			pubDate := it.Published
			if pubDate == "" {
				pubDate = it.Updated
			}
			cats := []string{}
			for _, c := range it.Categories {
				cats = append(cats, c.Term)
			}
			items = append(items, FeedItem{
				Title:       it.Title,
				Link:        it.Link.Href,
				PubDate:     pubDate,
				Description: it.Summary,
				Categories:  cats,
			})
		}
		return atom.Title, items
	}
	return "Без названия", []FeedItem{}
}

func (r *RssReader) addFeed(url, title string) Feed {
	if title == "" {
		content, err := r.fetchFeed(url)
		if err == nil {
			t, _ := r.parseFeed(content)
			title = t
		}
		if title == "" {
			title = url
		}
	}
	feed := Feed{
		ID:        r.nextFeedID,
		Title:     title,
		URL:       url,
		Items:     []FeedItem{},
		LastFetch: "",
	}
	r.nextFeedID++
	r.data.Feeds = append(r.data.Feeds, feed)
	r.save()
	return feed
}

func (r *RssReader) removeFeed(id int, url string) bool {
	for i, f := range r.data.Feeds {
		if (id > 0 && f.ID == id) || (url != "" && f.URL == url) {
			r.data.Feeds = append(r.data.Feeds[:i], r.data.Feeds[i+1:]...)
			r.save()
			return true
		}
	}
	return false
}

func (r *RssReader) listFeeds() []Feed {
	return r.data.Feeds
}

func (r *RssReader) fetchAll(filter Filter) []struct{ Feed string; Item FeedItem } {
	for i := range r.data.Feeds {
		feed := &r.data.Feeds[i]
		content, err := r.fetchFeed(feed.URL)
		if err != nil {
			fmt.Println(colorize("Ошибка при загрузке "+feed.URL+": "+err.Error(), red))
			continue
		}
		title, items := r.parseFeed(content)
		if title != feed.Title {
			feed.Title = title
		}
		for _, it := range items {
			exists := false
			for _, existing := range feed.Items {
				if existing.Link == it.Link {
					exists = true
					break
				}
			}
			if !exists {
				it.ID = r.nextItemID
				it.Read = false
				feed.Items = append(feed.Items, it)
				r.nextItemID++
			}
		}
		feed.LastFetch = time.Now().Format(time.RFC3339)
	}
	r.save()

	var all []struct{ Feed string; Item FeedItem }
	for _, feed := range r.data.Feeds {
		for _, item := range feed.Items {
			all = append(all, struct{ Feed string; Item FeedItem }{feed.Title, item})
		}
	}
	// Фильтрация
	filtered := r.applyFilters(all, filter)
	// Сортировка по дате
	for i := 0; i < len(filtered); i++ {
		for j := i + 1; j < len(filtered); j++ {
			if filtered[i].Item.PubDate < filtered[j].Item.PubDate {
				filtered[i], filtered[j] = filtered[j], filtered[i]
			}
		}
	}
	return filtered
}

func (r *RssReader) applyFilters(items []struct{ Feed string; Item FeedItem }, filter Filter) []struct{ Feed string; Item FeedItem } {
	var result []struct{ Feed string; Item FeedItem }
	for _, it := range items {
		// Ключевые слова
		if len(filter.Keywords) > 0 {
			text := strings.ToLower(it.Item.Title + " " + it.Item.Description + " " + strings.Join(it.Item.Categories, " "))
			match := false
			for _, kw := range filter.Keywords {
				if strings.Contains(text, strings.ToLower(kw)) {
					match = true
					break
				}
			}
			if !match {
				continue
			}
		}
		// Дата
		if filter.Days > 0 {
			pub, err := time.Parse(time.RFC3339, it.Item.PubDate)
			if err == nil {
				diff := time.Since(pub).Hours() / 24
				if diff > float64(filter.Days) {
					continue
				}
			}
		}
		// Источник
		if filter.Source != "" && !strings.Contains(strings.ToLower(it.Feed), strings.ToLower(filter.Source)) {
			continue
		}
		// Статус
		if filter.UnreadOnly && it.Item.Read {
			continue
		}
		if filter.ReadOnly && !it.Item.Read {
			continue
		}
		result = append(result, it)
	}
	return result
}

func (r *RssReader) getItem(id int) (string, *FeedItem) {
	for _, feed := range r.data.Feeds {
		for i := range feed.Items {
			if feed.Items[i].ID == id {
				return feed.Title, &feed.Items[i]
			}
		}
	}
	return "", nil
}

func (r *RssReader) markRead(id int) bool {
	_, item := r.getItem(id)
	if item != nil {
		item.Read = true
		r.save()
		return true
	}
	return false
}

func openLink(url string) {
	var cmd string
	switch os.Getenv("OS") {
	case "Windows_NT":
		cmd = "start"
	default:
		if _, err := exec.LookPath("xdg-open"); err == nil {
			cmd = "xdg-open"
		} else {
			cmd = "open"
		}
	}
	exec.Command(cmd, url).Run()
}

func (r *RssReader) exportOpml(filename string) {
	opml := `<?xml version="1.0" encoding="UTF-8"?>
<opml version="1.0">
<head><title>RSS Subscriptions</title></head>
<body>
`
	for _, feed := range r.data.Feeds {
		opml += fmt.Sprintf(`<outline text="%s" title="%s" type="rss" xmlUrl="%s"/>`+"\n", feed.Title, feed.Title, feed.URL)
	}
	opml += `</body>
</opml>`
	ioutil.WriteFile(filename, []byte(opml), 0644)
}

func (r *RssReader) importOpml(filename string) int {
	content, err := ioutil.ReadFile(filename)
	if err != nil {
		return 0
	}
	re := regexp.MustCompile(`<outline[^>]*xmlUrl="([^"]*)"[^>]*text="([^"]*)"[^>]*/?>`)
	matches := re.FindAllStringSubmatch(string(content), -1)
	count := 0
	for _, match := range matches {
		url := match[1]
		title := match[2]
		if title == "" {
			title = url
		}
		r.addFeed(url, title)
		count++
	}
	r.save()
	return count
}

func (r *RssReader) saveFilters() {
	r.save()
	fmt.Println(colorize("Фильтры сохранены", green))
}

func (r *RssReader) loadFilters() {
	// уже загружены
	fmt.Println(colorize("Фильтры загружены", green))
}

func (r *RssReader) clearFilters() {
	r.filter = Filter{}
	r.save()
	fmt.Println(colorize("Фильтры очищены", green))
}

func main() {
	var (
		cmd        string
		archive    string
		output     string
		keyword    string
		days       int
		source     string
		unread     bool
		read       bool
		clear      bool
		file       string
	)
	flag.StringVar(&cmd, "cmd", "", "Команда: add, remove, list, fetch, read, filter, save-filters, load-filters, export, import")
	flag.StringVar(&archive, "a", "", "Архив (не используется в RSS)")
	flag.StringVar(&output, "o", "", "Выходной файл")
	flag.StringVar(&keyword, "keyword", "", "Ключевое слово (можно несколько через запятую)")
	flag.IntVar(&days, "days", 0, "Показывать за последние N дней")
	flag.StringVar(&source, "source", "", "Фильтр по источнику")
	flag.BoolVar(&unread, "unread", false, "Только непрочитанные")
	flag.BoolVar(&read, "read", false, "Только прочитанные")
	flag.BoolVar(&clear, "clear-filters", false, "Очистить фильтры")
	flag.StringVar(&file, "file", "", "Файл для импорта/экспорта OPML")
	flag.Usage = func() {
		fmt.Println(`Использование: go run rss.go -cmd <команда> [опции]
  add       -url <url> -title <title>
  remove    -id <id> | -url <url>
  list
  fetch     [-keyword kw] [-days N] [-source src] [-unread] [-read]
  read      -id <id> [-text]
  filter    [-keyword kw] [-days N] [-source src] [-unread] [-read]
  save-filters
  load-filters
  export    -file <file>
  import    -file <file>`)
	}
	flag.Parse()

	if cmd == "" {
		fmt.Println("Укажите -cmd")
		os.Exit(1)
	}

	dataFile := "rss_data.json"
	reader := NewRssReader(dataFile)

	// Обработка фильтров
	if cmd == "filter" || keyword != "" || days > 0 || source != "" || unread || read || clear {
		if clear {
			reader.clearFilters()
			return
		}
		if keyword != "" {
			reader.filter.Keywords = strings.Split(keyword, ",")
		}
		if days > 0 {
			reader.filter.Days = days
		}
		if source != "" {
			reader.filter.Source = source
		}
		if unread {
			reader.filter.UnreadOnly = true
			reader.filter.ReadOnly = false
		}
		if read {
			reader.filter.ReadOnly = true
			reader.filter.UnreadOnly = false
		}
		reader.save()
		fmt.Println(colorize("Фильтры обновлены", green))
		return
	}

	if cmd == "save-filters" {
		reader.saveFilters()
		return
	}
	if cmd == "load-filters" {
		reader.loadFilters()
		return
	}

	switch cmd {
	case "add":
		url := flag.Arg(0)
		title := flag.Arg(1)
		if url == "" {
			fmt.Println("Требуется URL")
			os.Exit(1)
		}
		feed := reader.addFeed(url, title)
		fmt.Printf(colorize("Лента добавлена: %s (ID %d)\n", green), feed.Title, feed.ID)
	case "remove":
		id := 0
		fmt.Sscanf(flag.Arg(0), "%d", &id)
		url := flag.Arg(0)
		if id == 0 && url == "" {
			fmt.Println("Требуется --id или --url")
			os.Exit(1)
		}
		if reader.removeFeed(id, url) {
			fmt.Println(colorize("Лента удалена", green))
		} else {
			fmt.Println(colorize("Лента не найдена", red))
		}
	case "list":
		feeds := reader.listFeeds()
		if len(feeds) == 0 {
			fmt.Println("Нет лент.")
		} else {
			fmt.Println(colorize("Список лент:", bold+cyan))
			for _, f := range feeds {
				fmt.Printf("  %s: %s (%s)\n", colorize(strconv.Itoa(f.ID), green), f.Title, f.URL)
			}
		}
	case "fetch":
		items := reader.fetchAll(reader.filter)
		if len(items) == 0 {
			fmt.Println("Новостей нет (возможно, фильтры слишком строгие).")
		} else {
			fmt.Println(colorize("Новости (отфильтрованные):", bold+cyan))
			for _, it := range items {
				status := "⚪"
				if !it.Item.Read {
					status = "🔵"
				}
				date := it.Item.PubDate
				if len(date) > 16 {
					date = date[:16]
				}
				fmt.Printf("%s %s | %s | %s | %s\n",
					status,
					colorize(strconv.Itoa(it.Item.ID), green),
					colorize(truncate(it.Item.Title, 60), white),
					colorize(it.Feed, gray),
					colorize(date, yellow))
			}
		}
	case "read":
		idStr := flag.Arg(0)
		if idStr == "" {
			fmt.Println("Требуется --id")
			os.Exit(1)
		}
		id, _ := strconv.Atoi(idStr)
		feedTitle, item := reader.getItem(id)
		if item == nil {
			fmt.Println(colorize("Новость не найдена", red))
		} else {
			reader.markRead(id)
			fmt.Println(colorize("Заголовок: "+item.Title, bold+white))
			fmt.Println(colorize("Дата: "+item.PubDate, yellow))
			fmt.Println(colorize("Источник: "+feedTitle, gray))
			if flag.Lookup("text") != nil && flag.Lookup("text").Value.String() == "true" {
				fmt.Println(colorize("Содержание:", bold))
				fmt.Println(item.Description)
			} else {
				fmt.Println(colorize("Ссылка: "+item.Link, cyan))
				if item.Link != "" {
					openLink(item.Link)
					fmt.Println(colorize("Ссылка открыта в браузере", green))
				}
			}
		}
	case "export":
		filename := file
		if filename == "" {
			filename = "subscriptions.opml"
		}
		reader.exportOpml(filename)
		fmt.Println(colorize("Подписки экспортированы в "+filename, green))
	case "import":
		if file == "" {
			fmt.Println("Требуется --file")
			os.Exit(1)
		}
		count := reader.importOpml(file)
		fmt.Println(colorize(fmt.Sprintf("Импортировано %d подписок", count), green))
	default:
		fmt.Println("Неизвестная команда.")
	}
}

func truncate(s string, n int) string {
	if len(s) <= n {
		return s
	}
	return s[:n] + "..."
}

File: rss.php
<?php
// rss.php
// RSS Reader (Фильтрация новостей) на PHP

if (php_sapi_name() !== 'cli') {
    die("Это консольное приложение.\n");
}

// ANSI-цвета
define('RESET', "\033[0m");
define('BOLD', "\033[1m");
define('RED', "\033[91m");
define('GREEN', "\033[92m");
define('YELLOW', "\033[93m");
define('BLUE', "\033[94m");
define('MAGENTA', "\033[95m");
define('CYAN', "\033[96m");
define('WHITE', "\033[97m");
define('GRAY', "\033[90m");

function colorize($text, $color) {
    return $color . $text . RESET;
}

class Filter {
    public $keywords = [];
    public $days = 0;
    public $source = null;
    public $unreadOnly = false;
    public $readOnly = false;
}

class RssReader {
    private $dataFile;
    private $data;
    private $nextFeedId;
    private $nextItemId;
    public $filter;

    public function __construct($dataFile = 'rss_data.json') {
        $this->dataFile = $dataFile;
        $this->load();
    }

    private function load() {
        if (file_exists($this->dataFile)) {
            $json = file_get_contents($this->dataFile);
            $this->data = json_decode($json, true);
        }
        if (!$this->data) {
            $this->data = ['feeds' => [], 'next_feed_id' => 1, 'next_item_id' => 1];
        }
        $this->nextFeedId = $this->data['next_feed_id'] ?? 1;
        $this->nextItemId = $this->data['next_item_id'] ?? 1;
        $this->filter = new Filter();
        if (isset($this->data['filter'])) {
            $f = $this->data['filter'];
            $this->filter->keywords = $f['keywords'] ?? [];
            $this->filter->days = $f['days'] ?? 0;
            $this->filter->source = $f['source'] ?? null;
            $this->filter->unreadOnly = $f['unread_only'] ?? false;
            $this->filter->readOnly = $f['read_only'] ?? false;
        }
    }

    private function save() {
        $this->data['next_feed_id'] = $this->nextFeedId;
        $this->data['next_item_id'] = $this->nextItemId;
        $this->data['filter'] = [
            'keywords' => $this->filter->keywords,
            'days' => $this->filter->days,
            'source' => $this->filter->source,
            'unread_only' => $this->filter->unreadOnly,
            'read_only' => $this->filter->readOnly
        ];
        file_put_contents($this->dataFile, json_encode($this->data, JSON_PRETTY_PRINT | JSON_UNESCAPED_UNICODE));
    }

    private function fetchFeed($url) {
        $ctx = stream_context_create(['http' => ['timeout' => 10]]);
        return file_get_contents($url, false, $ctx);
    }

    private function parseFeed($content) {
        $title = 'Без названия';
        if (preg_match('/<title>(.*?)<\/title>/is', $content, $m)) $title = trim($m[1]);
        $items = [];
        preg_match_all('/<(item|entry)[^>]*>(.*?)<\/(item|entry)>/is', $content, $matches, PREG_SET_ORDER);
        foreach ($matches as $match) {
            $itemContent = $match[2];
            $item = [
                'title' => $this->extract($itemContent, 'title'),
                'link' => $this->extractLink($itemContent),
                'pubDate' => $this->extract($itemContent, 'pubDate|published|updated'),
                'description' => $this->extract($itemContent, 'description|summary|content'),
                'categories' => []
            ];
            preg_match_all('/<category[^>]*>([^<]*)<\/category>/i', $itemContent, $cats);
            if ($cats[1]) $item['categories'] = array_map('trim', $cats[1]);
            $items[] = $item;
        }
        return ['title' => $title, 'items' => $items];
    }

    private function extract($content, $tags) {
        if (preg_match('/<(' . $tags . ')>(.*?)<\/\\1>/is', $content, $m)) {
            return trim($m[2]);
        }
        return '';
    }

    private function extractLink($content) {
        if (preg_match('/<link>(.*?)<\/link>/is', $content, $m)) return trim($m[1]);
        if (preg_match('/<link[^>]*href="([^"]*)"/is', $content, $m)) return $m[1];
        return '';
    }

    public function addFeed($url, $title = null) {
        if (!$title) {
            try {
                $content = $this->fetchFeed($url);
                $parsed = $this->parseFeed($content);
                $title = $parsed['title'];
            } catch (Exception $e) {
                $title = $url;
            }
        }
        $feed = [
            'id' => $this->nextFeedId,
            'title' => $title,
            'url' => $url,
            'items' => [],
            'last_fetch' => null
        ];
        $this->data['feeds'][] = $feed;
        $this->nextFeedId++;
        $this->save();
        return $feed;
    }

    public function removeFeed($id, $url) {
        foreach ($this->data['feeds'] as $i => $f) {
            if (($id && $f['id'] == $id) || ($url && $f['url'] == $url)) {
                array_splice($this->data['feeds'], $i, 1);
                $this->save();
                return true;
            }
        }
        return false;
    }

    public function listFeeds() {
        return $this->data['feeds'];
    }

    public function fetchAll($filter = null) {
        if (!$filter) $filter = $this->filter;
        foreach ($this->data['feeds'] as &$feed) {
            try {
                $content = $this->fetchFeed($feed['url']);
                $parsed = $this->parseFeed($content);
                if ($parsed['title'] != $feed['title']) $feed['title'] = $parsed['title'];
                foreach ($parsed['items'] as $item) {
                    $exists = false;
                    foreach ($feed['items'] as $existing) {
                        if ($existing['link'] == $item['link']) { $exists = true; break; }
                    }
                    if (!$exists) {
                        $item['id'] = $this->nextItemId;
                        $item['read'] = false;
                        $feed['items'][] = $item;
                        $this->nextItemId++;
                    }
                }
                $feed['last_fetch'] = date('c');
            } catch (Exception $e) {
                echo colorize("Ошибка при загрузке {$feed['url']}: " . $e->getMessage(), RED) . "\n";
            }
        }
        $this->save();
        $all = [];
        foreach ($this->data['feeds'] as $f) {
            foreach ($f['items'] as $item) {
                $all[] = ['feed' => $f['title'], 'item' => $item];
            }
        }
        $all = $this->applyFilters($all, $filter);
        usort($all, function($a, $b) {
            return strcmp($b['item']['pubDate'], $a['item']['pubDate']);
        });
        return $all;
    }

    private function applyFilters($items, $filter) {
        $result = [];
        foreach ($items as $it) {
            $feedTitle = $it['feed'];
            $item = $it['item'];
            // Ключевые слова
            if (!empty($filter->keywords)) {
                $text = strtolower($item['title'] . ' ' . $item['description'] . ' ' . implode(' ', $item['categories']));
                $match = false;
                foreach ($filter->keywords as $kw) {
                    if (strpos($text, strtolower($kw)) !== false) { $match = true; break; }
                }
                if (!$match) continue;
            }
            // Дата
            if ($filter->days > 0) {
                try {
                    $pub = new DateTime($item['pubDate']);
                    $diff = (new DateTime())->diff($pub)->days;
                    if ($diff > $filter->days) continue;
                } catch (Exception $e) {}
            }
            // Источник
            if ($filter->source && stripos($feedTitle, $filter->source) === false) continue;
            // Статус
            if ($filter->unreadOnly && $item['read']) continue;
            if ($filter->readOnly && !$item['read']) continue;
            $result[] = $it;
        }
        return $result;
    }

    public function getItem($id) {
        foreach ($this->data['feeds'] as $f) {
            foreach ($f['items'] as $item) {
                if ($item['id'] == $id) {
                    return ['feedTitle' => $f['title'], 'item' => $item];
                }
            }
        }
        return null;
    }

    public function markRead($id) {
        $res = $this->getItem($id);
        if ($res) {
            foreach ($this->data['feeds'] as &$f) {
                foreach ($f['items'] as &$item) {
                    if ($item['id'] == $id) {
                        $item['read'] = true;
                        $this->save();
                        return true;
                    }
                }
            }
        }
        return false;
    }

    public function openLink($url) {
        if (PHP_OS_FAMILY === 'Windows') exec("start $url");
        elseif (PHP_OS_FAMILY === 'Darwin') exec("open $url");
        else exec("xdg-open $url");
    }

    public function exportOpml($filename = 'subscriptions.opml') {
        $opml = '<?xml version="1.0" encoding="UTF-8"?>' . "\n";
        $opml .= '<opml version="1.0">' . "\n";
        $opml .= '<head><title>RSS Subscriptions</title></head>' . "\n";
        $opml .= '<body>' . "\n";
        foreach ($this->data['feeds'] as $f) {
            $opml .= '<outline text="' . htmlspecialchars($f['title']) . '" title="' . htmlspecialchars($f['title']) . '" type="rss" xmlUrl="' . htmlspecialchars($f['url']) . '"/>' . "\n";
        }
        $opml .= '</body></opml>';
        file_put_contents($filename, $opml);
        return $filename;
    }

    public function importOpml($filename) {
        $content = file_get_contents($filename);
        preg_match_all('/<outline[^>]*xmlUrl="([^"]*)"[^>]*text="([^"]*)"[^>]*\/?>/i', $content, $matches, PREG_SET_ORDER);
        $count = 0;
        foreach ($matches as $m) {
            $url = $m[1];
            $title = $m[2] ?: $url;
            $this->addFeed($url, $title);
            $count++;
        }
        $this->save();
        return $count;
    }

    public function saveFilters() { $this->save(); echo colorize("Фильтры сохранены", GREEN) . "\n"; }
    public function loadFilters() { echo colorize("Фильтры загружены", GREEN) . "\n"; }
    public function clearFilters() { $this->filter = new Filter(); $this->save(); echo colorize("Фильтры очищены", GREEN) . "\n"; }
}

$args = array_slice($argv, 1);
if (empty($args) || $args[0] == 'help') {
    echo "Использование: php rss.php <команда> [опции]\n";
    echo "  add       --url <url> [--title <title>]\n";
    echo "  remove    --id <id> | --url <url>\n";
    echo "  list\n";
    echo "  fetch     [--keyword kw] [--days N] [--source src] [--unread] [--read]\n";
    echo "  read      --id <id> [--text]\n";
    echo "  filter    [--keyword kw] [--days N] [--source src] [--unread] [--read]\n";
    echo "  save-filters\n";
    echo "  load-filters\n";
    echo "  export    --file <file>\n";
    echo "  import    --file <file>\n";
    exit(0);
}
$command = $args[0];
$options = [];
for ($i = 1; $i < count($args); $i++) {
    if (strpos($args[$i], '--') === 0) {
        $key = substr($args[$i], 2);
        if (isset($args[$i+1]) && strpos($args[$i+1], '--') !== 0) {
            $options[$key] = $args[++$i];
        } else {
            $options[$key] = '';
        }
    }
}

$dataFile = $options['data'] ?? 'rss_data.json';
$reader = new RssReader($dataFile);

if ($command == 'filter' || isset($options['keyword']) || isset($options['days']) ||
    isset($options['source']) || isset($options['unread']) || isset($options['read'])) {
    if (isset($options['keyword'])) {
        $reader->filter->keywords = explode(',', $options['keyword']);
    }
    if (isset($options['days'])) $reader->filter->days = (int)$options['days'];
    if (isset($options['source'])) $reader->filter->source = $options['source'];
    if (isset($options['unread'])) { $reader->filter->unreadOnly = true; $reader->filter->readOnly = false; }
    if (isset($options['read'])) { $reader->filter->readOnly = true; $reader->filter->unreadOnly = false; }
    $reader->save();
    echo colorize("Фильтры обновлены", GREEN) . "\n";
    exit(0);
}

if ($command == 'save-filters') { $reader->saveFilters(); exit(0); }
if ($command == 'load-filters') { $reader->loadFilters(); exit(0); }

switch ($command) {
    case 'add':
        if (empty($options['url'])) { echo "Требуется --url\n"; exit(1); }
        $feed = $reader->addFeed($options['url'], $options['title'] ?? null);
        echo colorize("Лента добавлена: {$feed['title']} (ID {$feed['id']})", GREEN) . "\n";
        break;
    case 'remove':
        $id = isset($options['id']) ? (int)$options['id'] : 0;
        $url = $options['url'] ?? null;
        if ($id == 0 && !$url) { echo "Требуется --id или --url\n"; exit(1); }
        if ($reader->removeFeed($id, $url)) echo colorize("Лента удалена", GREEN) . "\n";
        else echo colorize("Лента не найдена", RED) . "\n";
        break;
    case 'list':
        $feeds = $reader->listFeeds();
        if (empty($feeds)) echo "Нет лент.\n";
        else {
            echo colorize("Список лент:", BOLD . CYAN) . "\n";
            foreach ($feeds as $f) {
                echo "  " . colorize($f['id'], GREEN) . ": {$f['title']} ({$f['url']})\n";
            }
        }
        break;
    case 'fetch':
        $items = $reader->fetchAll();
        if (empty($items)) echo "Новостей нет (возможно, фильтры слишком строгие).\n";
        else {
            echo colorize("Новости (отфильтрованные):", BOLD . CYAN) . "\n";
            foreach ($items as $it) {
                $status = $it['item']['read'] ? '⚪' : '🔵';
                $date = substr($it['item']['pubDate'] ?? '', 0, 16);
                echo "$status " . colorize($it['item']['id'], GREEN) . " | " . colorize(substr($it['item']['title'], 0, 60), WHITE) . " | " . colorize($it['feed'], GRAY) . " | " . colorize($date, YELLOW) . "\n";
            }
        }
        break;
    case 'read':
        if (empty($options['id'])) { echo "Требуется --id\n"; exit(1); }
        $id = (int)$options['id'];
        $res = $reader->getItem($id);
        if (!$res) echo colorize("Новость не найдена", RED) . "\n";
        else {
            $reader->markRead($id);
            echo colorize("Заголовок: " . $res['item']['title'], BOLD . WHITE) . "\n";
            echo colorize("Дата: " . $res['item']['pubDate'], YELLOW) . "\n";
            echo colorize("Источник: " . $res['feedTitle'], GRAY) . "\n";
            if (isset($options['text'])) {
                echo colorize("Содержание:", BOLD) . "\n";
                echo $res['item']['description'] . "\n";
            } else {
                echo colorize("Ссылка: " . $res['item']['link'], CYAN) . "\n";
                if ($res['item']['link']) {
                    $reader->openLink($res['item']['link']);
                    echo colorize("Ссылка открыта в браузере", GREEN) . "\n";
                }
            }
        }
        break;
    case 'export':
        $filename = $options['file'] ?? 'subscriptions.opml';
        $reader->exportOpml($filename);
        echo colorize("Подписки экспортированы в $filename", GREEN) . "\n";
        break;
    case 'import':
        if (empty($options['file'])) { echo "Требуется --file\n"; exit(1); }
        $count = $reader->importOpml($options['file']);
        echo colorize("Импортировано $count подписок", GREEN) . "\n";
        break;
    default:
        echo "Неизвестная команда.\n";
}
